reproducible/README.md

```diff
@@ -0,0 +1,343 @@
+         , ( "stdenv"
+           , "/nix/store/sm7kk5n84vaisqvhk1yfsjqls50j8s0m-stdenv-linux" )
+         , ("strictDeps", "")
+         , ("system", "x86_64-linux") ] )
+```
+
+The derivation can then be used to build the output using either
+```
+   nix-store --realize
+```
+or more compactly
+```
+   nix-build test-environment 
+```
+will perform both instantiate and realise. The result is a build output located in the nix-store.
+```
+/nix/store/syibqm72433pzxd46hyw3xbr5swq286j-test-environment
+```
+which contains:
+```
+$ ls /nix/store/syibqm72433pzxd46hyw3xbr5swq286j-test-environment
+qsgajw2yhrf33qgibsvd9y8zaw64w7j7-helloWorld
+```
+which in turn contains the helloWorld script.
+
+For more involved build project the build actions are clearly more complex, but the flow is the same.
+
+### The Nix Repo
+
+To build, a clone of the nix packages repo must be present. There is a default clone installed when nix 
+is installed. 
+
+The nix packages repo serves two purposes, somewhat at odds with each other.
+
+- The package repository for a linux distro "NixOS" which will eventully be a standalone distro. In that 
+  case packages are chosen for release branches with an eye to mutual compatibility and working well within
+  that defined universe.  
+
+- The package repository is also used for producing packages for other linux distros (ubuntu, debian,
+  redhat, fedora, arch, gentoo) with easily auditable dependencies. In this case, it is necessary to
+  consider specific compatible versions of packages.  In particular, glibc must be an equal or older
+  revision than the glibc of the platform or the program will not dynamically link.  This is by design.
+  Ubuntu 18.04 is limited to glibc 2.27. The current lts release 20.04 usues glibc 2.31 and fedora 33 
+  uses libc-2.32.  The nixpkgs branch release-20.09 is based on glibc 2.32.  The somewhat older branch
+  20.03 uses libc-2.30 so can work on the current common distros, including the current rhel.  
+
+  While it is not difficult to change the version of glibc in a custom repo branch, since glibc is linked 
+  by every package, to do so would require a custom repo build. In that case, it would require a custom 
+  cache to provide prebuilding, or everything would require everything be built from source..
+
+### Caching and Hydra
+
+The official branches in nixpkgs are prebuilt and available at cache.nixos.org. Caching is keyed to the build outputs
+sha digest, so custom environments will be able to use those prebuilts so long as no changes have been made. 
+If any changes are made to that package, any package related to the changed package will need to be custom
+built and cached in nix store.
+
+
+## A Nix build system for OpenEnclave SDK
+
+We propose to add an optional nix based build environment for OpenEnclave SDK using docker containers.
+
+This would consist of :
+- Dockerfile.nix and build-oe-nix-build.sh for building the oe nix docker image locally.
```

Comment:
I get `libsgx_enclave_common.so: no such file or directory` upon running `build-oe-nix-image.sh`:

```bash
$ sudo ./build-oe-nix-image.sh
...
...
+ nix-instantiate -I. -E '(import <nixpkgs> {}).dpkg'
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
/nix/store/dfdvxn42glnb9xgy8yrkd70gjnvhx6y0-dpkg-1.19.7.drv
+ nix-env -I. -i /nix/store/dfdvxn42glnb9xgy8yrkd70gjnvhx6y0-dpkg-1.19.7.drv
installing 'dpkg-1.19.7.drv'
these paths will be fetched (9.35 MiB download, 58.04 MiB unpacked):
  /nix/store/sg42gk32rbx5b3klmw9ql2sy165rnq8v-perl-5.30.1
  /nix/store/xic28d0cmcxv08wmh539xqq09vnajdar-dpkg-1.19.7
copying path '/nix/store/sg42gk32rbx5b3klmw9ql2sy165rnq8v-perl-5.30.1' from 'https://cache.nixos.org'...
copying path '/nix/store/xic28d0cmcxv08wmh539xqq09vnajdar-dpkg-1.19.7' from 'https://cache.nixos.org'...
building '/nix/store/9lc3laqsk26d3c95d38l2cfb716npa2j-user-environment.drv'...
created 742 symlinks in user environment
+ nix-env -I. -i /nix/store/6a6ilqysgz1gwfs0ahriw94q35vj84sy-vim-8.2.1123
installing 'vim-8.2.1123'
warning: unable to download 'https://cache.nixos.org/6a6ilqysgz1gwfs0ahriw94q35vj84sy.narinfo': Couldn't resolve host name (6); retrying in 260 ms
these paths will be fetched (14.82 MiB download, 70.98 MiB unpacked):
  /nix/store/0hcbvd7wpsx0p4vjvn5b2jb04rf35mkc-gawk-5.1.0
  /nix/store/2pi6zgkwnr3zdslvlv16nixpzvbyjx1n-glibc-2.31
  /nix/store/6a6ilqysgz1gwfs0ahriw94q35vj84sy-vim-8.2.1123
  /nix/store/j206pskg2yzyla1cnfrb9kb5n5bfvjgm-bash-4.4-p23
  /nix/store/mphihm7gn8yb1svf8zgyn9qvl86zhy6m-ncurses-6.2
  /nix/store/y7ya11l8qy5z2zyhc8db7km7kykkl1sw-libunistring-0.9.10
  /nix/store/zzp30ib8rgm42d80gw3jl5cy5vgvaq9y-libidn2-2.3.0
copying path '/nix/store/y7ya11l8qy5z2zyhc8db7km7kykkl1sw-libunistring-0.9.10' from 'https://cache.nixos.org'...
copying path '/nix/store/zzp30ib8rgm42d80gw3jl5cy5vgvaq9y-libidn2-2.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/2pi6zgkwnr3zdslvlv16nixpzvbyjx1n-glibc-2.31' from 'https://cache.nixos.org'...
copying path '/nix/store/j206pskg2yzyla1cnfrb9kb5n5bfvjgm-bash-4.4-p23' from 'https://cache.nixos.org'...
copying path '/nix/store/0hcbvd7wpsx0p4vjvn5b2jb04rf35mkc-gawk-5.1.0' from 'https://cache.nixos.org'...
copying path '/nix/store/mphihm7gn8yb1svf8zgyn9qvl86zhy6m-ncurses-6.2' from 'https://cache.nixos.org'...
copying path '/nix/store/6a6ilqysgz1gwfs0ahriw94q35vj84sy-vim-8.2.1123' from 'https://cache.nixos.org'...
building '/nix/store/j7icczmyfwb65nh6a0l7zha9j68dfysc-user-environment.drv'...
created 782 symlinks in user environment
+ nix-store -I. --verify --check-contents
error: no arguments expected
Try 'nix-store --help' for more information.
+ set +x
Removing intermediate container 482622a5e9af
 ---> de39efe358c0
Step 29/38 : ADD ./sort_deb_sum.sh /home/$BUILD_USER
 ---> 20b6a44fa5f3
Step 30/38 : ADD ./nix-build.sh /home/$BUILD_USER
 ---> e7a6fb1f6d1e
Step 31/38 : ADD ./nix-shell.sh /home/$BUILD_USER
 ---> c640511c64ac
Step 32/38 : ADD ./build_deb_pkg.sh /home/$BUILD_USER
 ---> 6c457a9e8e6a
Step 33/38 : RUN mkdir -p /home/$BUILD_USER/.nix_libs
 ---> Running in f84b6e911dc9
Removing intermediate container f84b6e911dc9
 ---> 86b48a4e1e37
Step 34/38 : ADD ./nix-libs/libsgx_enclave_common.so /home/$BUILD_USER/.nix_libs
ADD failed: stat /var/lib/docker/tmp/docker-builder901887648/nix-libs/libsgx_enclave_common.so: no such file or directory
...

```

Reply:
Investigating. I wonder if git has an issue with binaries

Reply:
Can we avoid checking in binaries?

Reply:
Just in case I'll put them into a tar file and copy from there.

Reply:
Ideally, we should host the tar file somewhere else (i.e., azure storage) and just use a script to download it.

Reply:
Is libsgx_enclave_common.so still needed to build the OE SDK binaries? Run time dependency on Intel SGX PSW components makes sense. Why do we have build time dependency?

Reply:
We don't need libsgx-enclave-common.so or any other sgx libraries to build the SDK.

Reply:
Not to build the sdk, but to run ctests.

Reply:
It would be good not to restrict visibility for now.  Its easy to retarget later.

Reply:
FWIW, if I have the tar file in the docker build directory, docker unpacks the contents. If I get it from https:// or whatever, I have to do it. Not an issue, the main question is where to find a CDN that is both open to everybody and secure.

Reply:
@anakrish @bodzhang responding to you both above. I am wondering what the testing strategy here is. I do have some thoughts. The package itself once built in the Nix environment could be exported and installed on the host and then verified with the sample installation. The alternative would be to build in nix, run in nix (ctest) and then export. The problem I foresee with the later methodology however is that the psw/dcap libs could be out of date, and if they are, attestation will not work due to not being able to generate a valid quote. 

What are your thoughts on a valid testing strategy, this can be brought up with SIG-Testing if we want to dig into it a little more offline but the concerns stem from a CI and Release standpoint as that is the end result here.

Reply:
@BRMcLaren , I also prefer to see the NIX build process limited to the build itself, and the ctest can be run on the build output. In fact, I wonder whether it's just luck that only a handful of the Intel PSW/DCAP .so files need to be included to run the current ctest. I would assume many other files are required to run any test that involves SGX attestation flow.

Reply:
@bodzhang I think we should be specific in 'build output' here, I believe we are both talking about the .deb produced, is something like this what you are imagining?

![image](https://user-images.githubusercontent.com/54328827/99863328-94482200-2b52-11eb-89fc-57d62261889c.png)

Reply:
@BRMcLaren , based on this [comment](https://github.com/openenclave/openenclave/pull/3744#discussion_r527970744), ctest runs before the .deb is produced. So If the intermediate "build output" can be used to run ctest outside the container, the test requirement is met. Of course, changing the build-ctest process a bit to output .deb and then run ctest works too.

Reply:
@bodzhang [currently](https://github.com/openenclave/openenclave/blob/master/.jenkins/pipelines/Azure/Linux/Jenkinsfile#L144-L168) is what we do. Build - > Test - > Package - > Test.

Awhile back we did just Build - > Test but ended up running into packaging issues, since then we always have stuck with the model of testing the package as well. Considering the end product of this is to have a highly auditable package it might be best to also do "changing the build-ctest process a bit to output .deb and then run ctest works too." up front.

Reply:
The Ctests are based on the cmake heirarchy, which in turn is oriented to the nix store paths.  I'm unaware of any process we have in place that can get around that.   Do we have a system for just installing the deb and running the unit tests on it?

Reply:
I've tried prototyping dumping the nix/store closure and reloading in another container with a similar nix store. That was able to run ctests directly. If the modle is we do the build, build the package then run ctests elsewhere, we can do that.

Reply:
To be clear, are we saying that this is fundamentally a parallel process, therefore we can take the success of the unit tests  for granted, build the deb, then sanity check by building/running the samples?

We can do either. We can build/ctest at once, or build then dump the nix/store closure, then run another container and ctest.  In all cases we can and should sanity check the deb by installing into a container and validating, which infrastructure I believe already exists

In all cases we should dump the closure for the sake of anyone who wants to stay in nix world to build their downstream project.

Reply:
So, you can't run ctest without the rest of the nix-store when you build in nix land, but you can export the closure (the build output plus any references made from the build output) then import it into a new instance of the container. I have done that, so not you build in one container and can test in the other.

A side effect of this is that it is possible to use the closure for openenclave-sdk to build additional downstream projects using the sdk. I will be adding that in a little bit.